server/src/api/commands.ts

```diff
@@ -1,49 +1,117 @@
 import * as os from 'os';
 
-const spawn = require('node-pty-prebuilt').spawn;
-
-interface CommandResult {
+interface CommandInformation {
+  id: string;
+  workspace: string;
   command: string;
   status: string;
+  outChunk: string;
   out: string;
+  factory: Function;
   commandRunning: any;
 }
 
-let commandRunIndex = 0;
-export let commandInProgress: CommandResult | null;
-
-export function runCommand(cwd: string, program: string, cmds: string[]) {
-  stopAllCommands();
-  const command = `${program} ${cmds.join(' ')} ${commandRunIndex++}`;
-  const commandRunning = spawn(program, cmds, { cwd, cols: 80 });
-  commandInProgress = {
-    command,
-    status: 'inprogress',
-    out: '',
-    commandRunning
-  };
-
-  commandRunning.on('data', (data: any) => {
-    if (commandInProgress && commandInProgress.command === command) {
-      commandInProgress.out += data.toString();
+export class RecentCommands {
+  commandInfos = [] as CommandInformation[];
+
+  constructor(private readonly MAX_COMMANDS: number) {}
+
+  addCommand(id: string, workspace: string, command: string, factory: any) {
+    if (this.commandInfos.length >= this.MAX_COMMANDS) {
+      if (!this.hasCompletedCommands(this.commandInfos)) {
+        throw new Error(
+          `Cannot run more than ${this.MAX_COMMANDS} commands in parallel`
+        );
+      }
+      this.commandInfos = this.withoutFirstCompletedCommand(this.commandInfos);
     }
-  });
 
-  commandRunning.on('exit', (code: any) => {
-    if (commandInProgress && commandInProgress.command === command) {
-      commandInProgress.status = code === 0 ? 'success' : 'failure';
+    this.commandInfos.push({
+      id,
+      workspace,
+      command,
+      status: 'waiting',
+      out: '',
+      outChunk: '',
+      factory,
+      commandRunning: null
+    });
+  }
+
+  restartCommand(id: string) {
+    const c = this.findMatchingCommand(id, this.commandInfos);
+    if (c.status === 'inprogress') {
+      this.stopCommand(id);
     }
-  });
-  return { command };
-}
+    c.status = 'inprogress';
+    c.commandRunning = c.factory();
+  }
+
+  addOut(id: string, out: string) {
+    const c = this.findMatchingCommand(id, this.commandInfos);
+    c.out += out;
+    c.outChunk += out;
+  }
+
+  setFinalStatus(id: string, status: string) {
```

Comment:
It feels like it would be nice to use an enum type here which the front end could use too. Could you maybe add a TODO here so we remember to clean that up later?